velox/exec/tests/SpillerTest.cpp

```diff
@@ -1395,6 +1424,67 @@ TEST_P(OrderByOutputOnly, basic) {
   }
 }
 
+class MaxSpillRunTest : public SpillerTest,
+                        public testing::WithParamInterface<TestParam> {
+ public:
+  MaxSpillRunTest() : SpillerTest(GetParam()) {}
+
+  static std::vector<TestParam> getTestParams() {
+    return TestParamsBuilder{.typesToExclude = {Spiller::Type::kHashJoinProbe}}
+        .getTestParams();
+  }
+};
+
+TEST_P(MaxSpillRunTest, basic) {
+  struct {
+    uint64_t maxSpillRunRows;
+    uint8_t expectedNumFiles;
+    std::string debugString() const {
+      return fmt::format(
+          "maxSpillRunRows {}, expectedNumFiles {}",
+          maxSpillRunRows,
+          expectedNumFiles);
+    }
+  } testSettings[] = {{0, 1}, {101, 3}, {4095, 3}, {4096, 3}, {4097, 2}};
+
+  auto numRows = 10'000;
+  for (const auto& testData : testSettings) {
+    SCOPED_TRACE(testData.debugString());
+    setupSpillData(rowType_, numKeys_, numRows, 1, nullptr, {});
+    sortSpillData();
+    setupSpiller(100'000, 0, 0, false, testData.maxSpillRunRows);
+    spiller_->spill();
+    ASSERT_FALSE(spiller_->finalized());
+    SpillPartitionSet spillPartitionSet;
+    spiller_->finishSpill(spillPartitionSet);
+    ASSERT_TRUE(spiller_->finalized());
+
+    auto numFiles{0};
+    auto totalSize{0};
+    std::vector<std::string_view> spilledFiles;
+    for (const auto& [_, sp] : spillPartitionSet) {
+      numFiles += sp->numFiles();
+      totalSize += sp->size();
+    }
+
+    const auto& stats = spiller_->stats();
+    ASSERT_EQ(totalSize, stats.spilledBytes);
+    if (type_ == Spiller::Type::kAggregateOutput ||
+        type_ == Spiller::Type::kOrderByOutput) {
+      ASSERT_EQ(numFiles, 1);
+      ASSERT_EQ(spillPartitionSet.size(), 1);
+    } else if (
+        type_ == Spiller::Type::kAggregateInput ||
+        type_ == Spiller::Type::kOrderByInput) {
+      // Need sort.
+      ASSERT_EQ(numFiles, testData.expectedNumFiles);
+      ASSERT_EQ(spillPartitionSet.size(), 1);
+    } else {
+      ASSERT_GT(numFiles, 1);
```

Comment:
If we set target file to an extreme large value, then the number of files will be 1 per each partition or at least smaller than the number of partitions? To verify that, we can have a test setup the number of partitions is one and four?

Reply:
Got it, I've made the target file size uint64_t max to force only one file per partition for the HashBuild spiller.
```C++
setupSpiller(std::numeric_limits<uint64_t>::max(),
   ...
} else {
      ASSERT_EQ(numFiles, hashBits_.numPartitions());
      ASSERT_EQ(spillPartitionSet.size(), numFiles);
    }
```